rust/sbp2json/src/lib.rs

```diff
@@ -2,18 +2,104 @@ use std::io::{Read, Write};
 
 use sbp::{
     json::{Json2JsonEncoder, JsonEncoder},
-    SbpEncoder,
+    HandleParseError, SbpEncoder,
 };
 use serde_json::ser::Formatter;
 
 pub type Result<T> = std::result::Result<T, Box<dyn std::error::Error + Send + Sync>>;
 
-pub fn json2sbp<R, W>(input: R, output: W, buffered: bool, fatal_errors: bool) -> Result<()>
+#[derive(Copy, Clone, Debug)]
+pub enum ErrorHandlerOptions {
+    ReturnOnFirstError,
+    FilterOutErrors,
+    CoerceErrorsToInvalidMsg,
+}
+
+impl std::str::FromStr for ErrorHandlerOptions {
+    type Err = String;
+
+    fn from_str(s: &str) -> std::result::Result<ErrorHandlerOptions, Self::Err> {
+        match s {
+            s if s.eq_ignore_ascii_case("Return") => Ok(ErrorHandlerOptions::ReturnOnFirstError),
+            s if s.eq_ignore_ascii_case("Skip") => Ok(ErrorHandlerOptions::FilterOutErrors),
+            s if s.eq_ignore_ascii_case("ToInvalid") | s.eq_ignore_ascii_case("to-invalid") => {
+                Ok(ErrorHandlerOptions::CoerceErrorsToInvalidMsg)
+            }
+
+            s => Err(format!(
+                "Unable to cast option {} to a valid error handler option. \
+                    Valid options are 'return', 'skip', & 'to-invalid')",
+                s
+            )),
+        }
+    }
+}
+
+/// Impl from bool for backwards compatibility
+impl From<bool> for ErrorHandlerOptions {
+    fn from(bool_flag: bool) -> Self {
+        if bool_flag {
+            Self::ReturnOnFirstError
+        } else {
+            Self::FilterOutErrors
+        }
+    }
+}
+
+/// This struct is a wrapper of an iter that returns messages or errors.
+pub struct ErrorHandler<'a, T, E: std::error::Error + Send + Sync + HandleParseError<T>> {
+    inner_iter: Box<dyn Iterator<Item = std::result::Result<T, E>> + 'a>,
+    opts: ErrorHandlerOptions,
+}
+
+impl<'a, T, E: std::error::Error + Send + Sync + HandleParseError<T>> ErrorHandler<'a, T, E> {
+    pub fn new(
+        inner_iter: impl Iterator<Item = std::result::Result<T, E>> + 'a,
+        opts: ErrorHandlerOptions,
+    ) -> Self {
+        Self {
+            inner_iter: Box::new(inner_iter),
+            opts,
+        }
+    }
+}
+
+impl<M, E: std::error::Error + Send + Sync + HandleParseError<M>> Iterator
+    for ErrorHandler<'_, M, E>
+{
+    type Item = M;
+
+    fn next(&mut self) -> Option<M> {
+        match (self.opts, self.inner_iter.next()) {
+            (_, None) => None,
+            (_, Some(Ok(x))) => Some(x),
+            (ErrorHandlerOptions::ReturnOnFirstError, Some(Err(e))) => {
+                eprintln!("{e}");
```

Comment:
Are these eprintln's here for debugging purposes, do they need remain here long term?

Reply:
It looks like this maybe replaces `maybe_fatal_errors`?

Reply:
I was thinking this might be a global impl, but since it's limited to the sbp2json library then have `eprintln` in the implementation probably isn't an issue.

Reply:
the `eprintln!` here was basically just maintaining the same behavior from `maybe_fatal_errors!` as it prints all errors to stderr. I don't believe i added any print statements in the lib but i will check now.
```rust
fn maybe_fatal_errors<'a, M, I, E>(
    messages: I,
    fatal_errors: bool,
) -> Box<dyn Iterator<Item = M> + 'a>
where
    M: 'a,
    I: Iterator<Item = std::result::Result<M, E>> + 'a,
    E: std::error::Error + 'a,
{
    if fatal_errors {
        Box::new(
            messages
                .take_while(|m| m.as_ref().map_err(|e| eprintln!("{e}")).is_ok())
                .map(|m| m.unwrap()),
        )
    } else {
        Box::new(messages.filter_map(|m| m.map_err(|e| eprintln!("{e}")).ok()))
    }
}
```